datafusion/physical-expr/src/utils/guarantee.rs

```diff
@@ -0,0 +1,518 @@
+///
+/// See [`LiteralGuarantee::analyze`] to extract literal guarantees from a
+/// predicate.
+///
+/// # Details
+/// A guarantee can be one of two forms:
+///
+/// 1. One of particular set of values. For example, `(a = 1)`, `(a = 1 OR a =
+/// 2) or `a IN (1, 2, 3)`
+///
+/// 2. NOT one of a particular set of values. For example, `(a != 1)`, `(a != 1
+/// AND a != 2)` or `a NOT IN (1, 2, 3)`
+///
+#[derive(Debug, Clone, PartialEq)]
+pub struct LiteralGuarantee {
+    pub column: Column,
+    pub guarantee: Guarantee,
+    pub literals: HashSet<ScalarValue>,
+}
+
+/// What is be guaranteed about the values?
+#[derive(Debug, Clone, PartialEq)]
+pub enum Guarantee {
+    /// `column` is one of a set of constant values
+    In,
+    /// `column` is NOT one of a set of constant values
+    NotIn,
+}
+
+impl LiteralGuarantee {
+    /// Create a new instance of the guarantee if the provided operator is
+    /// supported. Returns None otherwise. See [`LiteralGuarantee::analyze`] to
+    /// create these structures from an expression.
+    fn try_new<'a>(
+        column_name: impl Into<String>,
+        op: &Operator,
+        literals: impl IntoIterator<Item = &'a ScalarValue>,
+    ) -> Option<Self> {
+        let guarantee = match op {
+            Operator::Eq => Guarantee::In,
+            Operator::NotEq => Guarantee::NotIn,
+            _ => return None,
+        };
+
+        let literals: HashSet<_> = literals.into_iter().cloned().collect();
+
+        Some(Self {
+            column: Column::from_name(column_name),
+            guarantee,
+            literals,
+        })
+    }
+
+    /// Return a list of `LiteralGuarantees` for the specified predicate that
+    /// hold if the predicate (filter) expression evaluates to `true`.
+    ///
+    /// `expr` should be a boolean expression
+    ///
+    /// Note: this function does not simplify the expression prior to analysis.
+    pub fn analyze(expr: &Arc<dyn PhysicalExpr>) -> Vec<LiteralGuarantee> {
+        split_conjunction(expr)
+            .into_iter()
+            .fold(GuaranteeBuilder::new(), |builder, expr| {
+                if let Some(cel) = ColOpLit::try_new(expr) {
+                    return builder.aggregate_conjunct(cel);
+                } else {
+                    // look for (col <op> literal) OR (col <op> literal) ...
+                    let disjunctions = split_disjunction(expr);
+
+                    let terms = disjunctions
+                        .iter()
+                        .filter_map(|expr| ColOpLit::try_new(expr))
+                        .collect::<Vec<_>>();
+
+                    if terms.is_empty() {
+                        return builder;
+                    }
+
+                    // if not all terms are of the form (col <op> literal),
+                    // can't infer anything
+                    if terms.len() != disjunctions.len() {
+                        return builder;
+                    }
```

Comment:
Why is this check necessary? We can infer the guarantee for expr like below,  as dropping a part of OR-chain the condition still holds
```
... AND (
  a != 1 OR
  b > 1 OR
  a != 2
)
```

Reply:
That is an excellent point @waynexia. In fact I think this logic could potentially be improved even more

For example `a = 1 OR  a != 2` could actually provide two guarantees that could prove the predicate was false
1. `a != 1`
2. `a = 2`

Reply:
I thought a lot about this last night and this morning -- at the core what this code is trying to do is help prove an expression can not be `true` (so rows can be filtered)

To do so, it is trying to provide one of the following guarantees:
1. For the expression to be `true`, a column can **ONLY** be one of a set of constants
2. For the expression to be `true` a column can **NOT BE ANY** of a set of constants

For the expression 
```
  a != 1 OR
  b > 1 OR
  a != 2
```

We can't prove that the expression is true **ONLY** if a is not in `(1, 2)` -- it could be true if b was greater than 1 as well.

This subtelty was not clear from the comments so I have:
1. Updated the comments
2. Added more tests that illustrate this type of situation

I am pretty sure the new tests have uncovered a bug which I am working on fixing .

Thanks again for raising this point

Reply:
I plan to look at this tomorrow closely

Reply:
Thank k you -- I plan to have updated the code again by that time